File: LeetCode/bootleg/search_for_a_range.cs
//  https://leetcode.com/problems/search-for-a-range/
//
//  Given a sorted array of integers, find the starting and ending position of a given target value.
//
//  Your algorithm's runtime complexity must be in the order of O(log n).
//
//  If the target is not found in the array, return [-1, -1].
//
//  For example,
//  Given [5, 7, 7, 8, 8, 10] and target value 8,
//  return [3, 4]. 

using System;

public class Solution {
    //  
    //  Submission Details
    //  81 / 81 test cases passed.
    //      Status: Accepted
    //      Runtime: 496 ms
    //          
    //          Submitted: 0 minutes ago
    //
    //  https://leetcode.com/submissions/detail/51687811/
    //
    //  You are here!
    //  Your runtime beats 69.49% of csharp submissions.
    public int[] SearchRange(int[] nums, int target) {
        var result = new int[2];
        result[0] = Search(nums, target, left: true);
        if (result[0] == -1)
        {
            result[1] = -1;
            return result;
        }

        result[1] = Search(nums, target, left: false);
        return result;
    }

    public int Search(int[] nums, int target, bool left)
    {
        var low = 0;
        var high = nums.Length - 1;
        var index = -1;
        while (low <= high)
        {
            var mid = low + (high - low) / 2;
            if (nums[mid] == target)
            {
                index = mid;
                if (left)
                {
                    high = mid - 1;
                }
                else
                {
                    low = mid + 1;
                }

                continue;
            }

            if (nums[mid] < target)
            {
                low = mid + 1;
            }
            else
            {
                high = mid - 1;
            }
        }

        return index;
    }

    static void Main()
    {
        Console.WriteLine(String.Format("[{0}]", String.Join(", ", new Solution().SearchRange(new [] { 5, 7, 7, 8, 8, 10 }, 8))));
        Console.WriteLine(String.Format("[{0}]", String.Join(", ", new Solution().SearchRange(new [] { 1 }, 1))));
    }
}

File: LeetCode/remote/search_for_a_range.rb
#   https://leetcode.com/problems/search-for-a-range/
#
#   Given a sorted array of integers, find the starting and ending position of a given target value.
#
#   Your algorithm's runtime complexity must be in the order of O(log n).
#
#   If the target is not found in the array, return [-1, -1].
#
#   https://leetcode.com/submissions/detail/54795842/
#
#   Submission Details
#   81 / 81 test cases passed.
#       Status: Accepted
#       Runtime: 80 ms
#           
#           Submitted: 0 minutes ago

def search_range(nums, target)
    left = search(target, nums, :left)
    return [-1, -1] if left == -1
    [left, search(target, nums, :right)]
end

def search(value, nums, direction)
    low = 0
    high = nums.size
    found = -1
    while low < high
        mid = low + (high - low) / 2
        if nums[mid] > value
            high = mid
        elsif nums[mid] < value
            low = mid + 1
        else
            found = mid
            case direction
                when :left  
                    high = mid
                when :right 
                    low = mid + 1
            end
        end
    end
    
    found
end

# Way nicer solution
#   https://leetcode.com/submissions/detail/54795980/
#
#   Submission Details
#   81 / 81 test cases passed.
#       Status: Accepted
#       Runtime: 72 ms
#           
#           Submitted: 0 minutes ago
def search_range(nums, target)
    left = search(target, nums, :left)
    return [-1, -1] if left == nums.size || target != nums[left]
    [left, search(target + 1, nums, :right) - 1]
end

def search(value, nums, direction)
    low = 0
    high = nums.size
    while low < high
        mid = low + (high - low) / 2
        if nums[mid] < value
            low = mid + 1
        else
            high = mid
        end
    end
    
    low
end

#   It's evolution baby!
#   
#   Submission Details
#   81 / 81 test cases passed.
#       Status: Accepted
#       Runtime: 92 ms
#           
#           Submitted: 0 minutes ago
#
def search_range(nums, target)
    low = search(nums, target)
    nums[low] == target ? [low, search(nums, target + 1) - 1] : [-1, -1] 
end

def search(nums, target)
    low = 0
    high = nums.size
    while low < high
        mid = low + (high - low) / 2
        if nums[mid] < target
            low = mid + 1
        else
            high = mid
        end
    end
    
    return low
end
